# HG changeset patch
# User Betterbird <betterbird@betterbird.eu>
# Date 1652906465 -7200
# Parent  5d1716ed474b28fcfe6ac46073f966150b78ac11
Feature: Add "Open Message For ID" to context menu in news message.
https://bugzilla.mozilla.org/show_bug.cgi?id=37653 from the year 2000.
* * *
Feature: Restrict 'Open Message For ID' to message pane for now. Doesn't work for message in tab.

diff --git a/mail/base/content/mailContextMenus.js b/mail/base/content/mailContextMenus.js
--- a/mail/base/content/mailContextMenus.js
+++ b/mail/base/content/mailContextMenus.js
@@ -625,16 +625,20 @@ function IsMenuItemShowing(menuID) {
 function addEmail(url = gContextMenu.linkURL) {
   let addresses = getEmail(url);
   toAddressBook({
     action: "create",
     address: addresses,
   });
 }
 
+function openMessageForMsgId() {
+  OpenMessageForMessageId(getEmail(gContextMenu.linkURL));
+}
+
 function composeEmailTo(linkURL, identity) {
   let fields = Cc[
     "@mozilla.org/messengercompose/composefields;1"
   ].createInstance(Ci.nsIMsgCompFields);
   let params = Cc[
     "@mozilla.org/messengercompose/composeparams;1"
   ].createInstance(Ci.nsIMsgComposeParams);
   fields.to = getEmail(linkURL);
diff --git a/mail/base/content/mainPopupSet.inc.xhtml b/mail/base/content/mainPopupSet.inc.xhtml
--- a/mail/base/content/mainPopupSet.inc.xhtml
+++ b/mail/base/content/mainPopupSet.inc.xhtml
@@ -406,16 +406,22 @@
     <menuitem id="mailContext-copylink"
               label="&copyLinkCmd.label;"
               accesskey="&copyLinkCmd.accesskey;"
               command="cmd_copyLink"/>
     <menuitem id="mailContext-copyimage"
               label="&copyImageAllCmd.label;"
               accesskey="&copyImageAllCmd.accesskey;"
               command="cmd_copyImage"/>
+#ifndef BROWSER_POPUPS_ONLY
+    <menuitem id="mailContext-openMessageForMsgId"
+              label="&OpenMessageForMsgId.label;"
+              accesskey="&OpenMessageForMsgId.accesskey;"
+              oncommand="openMessageForMsgId();"/>
+#endif
     <menuitem id="mailContext-addemail"
               label="&AddToAddressBook.label;"
               accesskey="&AddToAddressBook.accesskey;"
               oncommand="addEmail();"/>
     <menuitem id="mailContext-composeemailto"
               label="&SendMessageTo.label;"
               accesskey="&SendMessageTo.accesskey;"
               oncommand="composeEmailTo(gContextMenu.linkURL);"/>
diff --git a/mail/base/content/nsContextMenu.js b/mail/base/content/nsContextMenu.js
--- a/mail/base/content/nsContextMenu.js
+++ b/mail/base/content/nsContextMenu.js
@@ -427,16 +427,24 @@ class nsContextMenu {
       "mailContext-selectall",
       !this.inThreadPane && !this.onPlayableMedia
     );
     this.showItem("mailContext-copyemail", this.onMailtoLink);
     this.showItem("mailContext-copylink", this.onLink && !this.onMailtoLink);
     this.showItem("mailContext-copyimage", this.onImage);
 
     this.showItem(
+      "mailContext-openMessageForMsgId",
+      this.onMailtoLink &&
+        !this.inThreadPane &&
+        this.isNewsgroup &&
+        // This fails badly if message opened in tab.
+        gFolderDisplay._folderPaneVisible
+    );
+    this.showItem(
       "mailContext-composeemailto",
       this.onMailtoLink && !this.inThreadPane
     );
     this.showItem(
       "mailContext-addemail",
       this.onMailtoLink && !this.inThreadPane
     );
 
